snapcraft/internal/lifecycle.py

```diff
@@ -112,31 +112,34 @@ def __init__(self, config, project_options):
         self.config = config
         self.project_options = project_options
         self.parts_config = config.parts
+        self._steps_run = {p: set() for p in self.config.part_names}
 
-    def run(self, step, part_names=None, recursed=False):
+    def run(self, step, part_names=None):
         if part_names:
             self.parts_config.validate(part_names)
-            parts = {p for p in self.config.all_parts if p.name in part_names}
+            parts = [p for p in self.config.all_parts if p.name in part_names]
         else:
             parts = self.config.all_parts
             part_names = self.config.part_names
 
-        dirty = {p.name for p in parts if p.should_step_run('stage')}
         step_index = common.COMMAND_ORDER.index(step) + 1
 
         for step in common.COMMAND_ORDER[0:step_index]:
             if step == 'stage':
                 pluginhandler.check_for_collisions(self.config.all_parts)
             for part in parts:
-                self._run_step(step, part, part_names, dirty, recursed)
+                self._run_step(step, part, part_names)
 
         self._create_meta(step, part_names)
 
-    def _run_step(self, step, part, part_names, dirty, recursed):
+    def _run_step(self, step, part, part_names):
+        if step in self._steps_run[part.name]:
+            return
+
         common.reset_env()
         prereqs = self.parts_config.get_prereqs(part.name)
-        if recursed:
-            prereqs = prereqs & dirty
+        prereqs = {p for p in prereqs if 'stage' not in self._steps_run[p]}
```

Comment:
El 13/09/16 a las 11:26, Kyle Fazzari escribió:
>
> In snapcraft/internal/lifecycle.py 
> <https://github.com/snapcore/snapcraft/pull/794#discussion_r78570244>:
>
> >          common.reset_env()
> >          prereqs = self.parts_config.get_prereqs(part.name)
> > -        if recursed:
> > -            prereqs = prereqs & dirty
> > +        prereqs = {p for p in prereqs if 'stage' not in self._steps_run[p]}
>
> My comment about moving this into an |if| below still applies.
>
But I am making the ifs below easier to handle by doing this here. 
Creating a set with prereqs that need to be staged still.